### PR TITLE
ajour: init at 0.6.3

### PR DESCRIPTION
**Pull request by mweinelt**
###### Motivation for this change

A cross-plattform addon manager for World of Warcraft.

https://github.com/casperstorm/ajour

cc @MetaDark @cawilliamson

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mweinelt** [comment]
It starts up, looks pretty, but I can't select a game directory just yet.

**Comment by abathur** [comment]
Off-topic: You may already know this, but I guess it's a recent-ish addition that I only noticed this fall myself, and the peace-of-mind it brings is nice.

There's an option hiding in a dropdown to the right of the Create pull request button--and then hidden away under the reviewers section of the PR info--for marking it as draft:

Screen Shot 2021-01-07 at 7 12 02 PM

Screen Shot 2021-01-07 at 7 03 49 PM

**Comment by mweinelt** [comment]
Thanks, marked as draft, and your post as off-topic, as you said yourself.

File: pkgs/tools/games/ajour/default.nix [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+, pkg-config
+, expat
+, openssl
+, libxkbcommon
+, vulkan-headers
+, vulkan-loader
+, wayland
+, x11
+, xorg
+, python3
```
Comment (MetaDark) [inline pkgs/tools/games/ajour/default.nix:15]: personal preference nitpick: input order should roughly reflect usage order

File: pkgs/tools/games/ajour/default.nix [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+}:
+
+let
+  rpathLibs = [
+    libxkbcommon
+    vulkan-loader
+    wayland
+  ];
```
Comment (MetaDark) [inline pkgs/tools/games/ajour/default.nix:23]: I noticed that X11 wasn't listed as a load time dependency through ldd, so I switched over to sway and found that these also need to be added to `rpathLibs` for X11 support:
```suggestion
  rpathLibs = [
    libXcursor
    libXi
    libxkbcommon
    libXrandr
    vulkan-loader
    wayland
    x11
  ];
```

**Review by kira-bruneau — CHANGES_REQUESTED** [review]
Thanks @mweinelt! I don't actually play WoW (even though I still have it installed :stuck_out_tongue:), but I can still help review this.

Would the game directory selection normally open a GTK / Qt file dialog? You might need to specify some of the environment variables that are normally set in `wrapGAppsHook` or `wrapQtAppsHook`.

**Review by kira-bruneau — CHANGES_REQUESTED** [review]

**Comment by MetaDark** [comment]
@mweinelt It looks like `native-dialog` requires either zenity or kdialog to be in `PATH`: https://github.com/balthild/native-dialog-rs/blob/master/src/dialog_impl/gnu/mod.rs.

File: pkgs/tools/games/ajour/default.nix [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  meta = with lib; {
+    description = "World of Warcraft addon manager written in Rust";
+    longDescription = ''
+      Ajour is a World of Warcraft addon manager written in Rust with a
+      strong focus on performance and simplicity. The project is
+      completely advertisement free, privacy respecting and open source.
+    '';
+    homepage = "https://github.com/casperstorm/ajour";
+    changelog = "https://github.com/casperstorm/ajour/blob/master/CHANGELOG.md";
+    licenses = license.mit;
```
Comment (SuperSandro2000) [inline pkgs/tools/games/ajour/default.nix:65]:
```suggestion
    license = license.mit;
```

Comment (MetaDark) [reply pkgs/tools/games/ajour/default.nix:65]: Nice catch! It also looks like `license.mit` should be changed to `licenses.mit`. It seems like the words were just swapped.

File: pkgs/tools/games/ajour/default.nix (hunk 2 above)
Comment (mweinelt) [reply pkgs/tools/games/ajour/default.nix]: I indeed only tested this on wayland at this point. Thanks for coming up with the list for X11.

File: pkgs/tools/games/ajour/default.nix [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    pkg-config
+    python3
+  ];
+
+  buildInputs = [
+    expat
+    openssl
+    xorg.libxcb
+    x11
+  ];
```
[the remark this answers is not in the document]
Comment (mweinelt) [reply pkgs/tools/games/ajour/default.nix]: Thanks for the clarification.

**Comment by mweinelt** [comment]
> Would the game directory selection normally open a GTK / Qt file dialog? You might need to specify some of the environment variables that are normally set in `wrapGAppsHook` or `wrapQtAppsHook`.

Yeah, that was it. I went ahead and put zenity and kdialog in the PATH and it is working now.

**Review by kira-bruneau — APPROVED** [review]

**Comment by MetaDark** [comment]
Result of `nixpkgs-review pr 108738` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ajour</li>
  </ul>
</details>

**Comment by piegamesde** [comment]
When I run the package on Arch Linux, it simply does nothing and exits with code 101 right away :(

The strace line `write(3, "02:25:59.712 [panic][ERROR] thre"..., 453) = 453` suggests that the code aborts during a panic.

**Comment by MetaDark** [comment]
@piegamesde Do you know if this panic is caused by a missing runtime library? You can check with `LD_DEBUG=libs ajour`

**Comment by piegamesde** [comment]
So, apparently exit code 101 is the panic exit code. So if this is not an abort, why am I not seeing a backtrace?

I don't know how `LD_DEBUG=libs` output looks like on a failure, but I can't see any obviously failing things there.

**Comment by MetaDark** [comment]
@piegamesde It looks like ajour uses [log-panics](https://docs.rs/log-panics/2.0.0/log_panics/) to capture panic messages to a log file. Looking through the source, it seems like it should be writing them to `~/.config/ajour/ajour.log`.

**Comment by piegamesde** [comment]
@MetaDark Thanks. That explains why the strace command showed the application writing to FD 3 instead of 2 … yeah, there's definitely something of interest in there:

```
02:38:15.342 [ajour][INFO] Ajour 0.6.1 has started.
02:38:15.342 [ajour::gui][DEBUG] config loaded:
Config {
    wow: Wow {
        directory: None,
        flavor: Retail,
    },
    addons: Addons {
        global_release_channel: Stable,
        ignored: {},
        release_channels: {},
        delete_saved_variables: false,
    },
    theme: None,
    column_config: V1 {
        local_version_width: 150,
        remote_version_width: 150,
        status_width: 85,
    },
    window_size: None,
    scale: None,
    backup_directory: None,
    backup_addons: false,
    backup_wtf: false,
    hide_ignored_addons: false,
    self_update_channel: Stable,
    weak_auras_account: {},
    alternating_row_colors: false,
}
02:38:15.342 [ajour::gui][DEBUG] antialiasing: true
02:38:15.358 [panic][ERROR] thread 'main' panicked at 'running Ajour gui: GraphicsAdapterNotFound': src/gui/mod.rs:1053
   0: log_panics::init::{{closure}}
   1: std::panicking::rust_panic_with_hook
   2: rust_begin_unwind
   3: core::panicking::panic_fmt
   4: core::result::unwrap_failed
   5: ajour::gui::run
   6: ajour::main
   7: std::rt::lang_start::{{closure}}
   8: std::rt::lang_start_internal
   9: main
  10: __libc_start_main
  11: _start
```

**Comment by mweinelt** [comment]
> 02:38:15.358 [panic][ERROR] thread 'main' panicked at 'running Ajour gui: GraphicsAdapterNotFound': src/gui/mod.rs:1053

Funny, that's basically how I started out, before I started patching the RPATH. Can you test the latest change? It adds all the libraries that I set the RPATH for to buildInputs.

**Comment by piegamesde** [comment]
Nope, no change :/ maybe I should open a bug upstream asking for better log diagnostics …

**Comment by mweinelt** [comment]
@piegamesde Can you strace the process and check if it is unable to open any libraries?

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 108738` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>ajour</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 108738` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>ajour</li>
</ul>
</details>

**Comment by MetaDark** [comment]
Result of `nixpkgs-review pr 108738` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ajour</li>
  </ul>
</details>

**Review by kira-bruneau — APPROVED** [review]

**Comment by mweinelt** [comment]
Something is really weird.

> anonymous function at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/pkgs/tools/games/ajour/default.nix:1:1 called without required argument 'kdialog', at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/lib/customisation.nix:69:16

```
❯ nix-build -A kdialog
/nix/store/v1gs4i0akskajmin3kvid5pvxa15zkda-kdialog-20.08.3
```

I renamed this from `kdeApplications.kdialog` as of a few days ago, as I couldn't find that anymore, and still can't.

```
❯ nix-build -A kdeApplication.kdialog
error: --- AttrPathNotFound ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------- nix-build
attribute 'kdeApplication' in selection path 'kdeApplication.kdialog' not found
```

I've just rebased this onto master and it still builds for me with just `kdialog`. I am clueless.

**Comment by piegamesde** [comment]
The kdeApplications problem was introduced by #102168

**Comment by mweinelt** [comment]
Migrated to `plasma5Packages.kdialog. Thanks for looking this up.

File: pkgs/tools/games/ajour/default.nix [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+, rustPlatform
+, autoPatchelfHook
+, cmake
+, makeWrapper
+, pkg-config
+, python3
+, expat
+, freetype
+, gnome3
+, plasma5Packages
```
Comment (SuperSandro2000) [inline pkgs/tools/games/ajour/default.nix:12]: Please don't input plasma5Packages. It should be inherit from all-packages.

File: pkgs/tools/games/ajour/default.nix [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+, fetchFromGitHub
+, rustPlatform
+, autoPatchelfHook
+, cmake
+, makeWrapper
+, pkg-config
+, python3
+, expat
+, freetype
+, gnome3
```
Comment (SuperSandro2000) [inline pkgs/tools/games/ajour/default.nix:11]: Same here.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/tools/games/ajour/default.nix (hunk 5 above)
Comment (mweinelt) [reply pkgs/tools/games/ajour/default.nix:12]: Can you reference some docs? Why is it fine for gnome3?

Comment (SuperSandro2000) [reply pkgs/tools/games/ajour/default.nix]: I saw that on a recent comment from ttuegel but I probably won't find it.

**Pull request merged**